{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Topic Modeling with SARS-Cov-2 Genome ðŸ§¬\n",
    "=======================================\n",
    "\n",
    "Problem description: SARS-CoV-2 is spreading across the world and as it\n",
    "spreads mutations are occuring. A way to understand the spreading and\n",
    "the mutations is to explore the strucure and information hidden in the\n",
    "genome.\n",
    "\n",
    "Project goal: The Goal of this project is to explore a SARS-CoV-2 genome\n",
    "dataset and try to predict the origin of a SARS-CoV-2 genome sample.\n",
    "\n",
    "Data: We will use publicly available NCBI SARS-CoV-2 genome with their\n",
    "geographic region information.\n",
    "\n",
    "| Geographic region | \\#samples | |:-----------------: |:--------: | |\n",
    "Africa | 397 | | Asia | 2534 | | Europe | 1418 | | North America | 26836\n",
    "| | Oceania | 13304 | | South America | 158 |\n",
    "\n",
    "Data link:\n",
    "https://www.ncbi.nlm.nih.gov/labs/virus/vssi/\\#/virus?SeqType*s=Nucleotide&VirusLineage*ss=Severe%20acute%20respiratory%20syndrome%20coronavirus%202%20(SARS-CoV-2),%20taxid:2697049\n",
    "\n",
    "#### Background:\n",
    "\n",
    "-   k-mer: https://en.wikipedia.org/wiki/K-mer\n",
    "-   genome: https://en.wikipedia.org/wiki/Genome\n",
    "-   Sequence analysis of SARS-CoV-2 genome reveals features important\n",
    "    for vaccine design:\n",
    "    https://www.nature.com/articles/s41598-020-72533-2\n",
    "-   Latent Dirichlet Allocation (LDA) tutorial from the course\n",
    "\n",
    "#### Challenges:\n",
    "\n",
    "-   How to encode genome sequence?\n",
    "    -   *Project solution:* Extract features with Latent Dirichlet\n",
    "        Allocation (LDA) by considering each genome sequence as a\n",
    "        document and each 3-mer as a word. So, we have a collection of\n",
    "        genomes consisting of 3-mers.\n",
    "-   How to relate encoded features to the origins?\n",
    "    -   *Project solution:* We used a Random Forest Classifier. One\n",
    "        advantage is interpretability: we can understand the positive or\n",
    "        negative relations a topic has on the origin.\n",
    "-   How to solve unbalanced class problem? E.g. North America has 26836\n",
    "    samples but South America has only 158\n",
    "    -   *Project solution:* Use f1 measure as metric\n",
    "\n",
    "Project steps: 1. Get SARS-CoV-2 data from NCBI 1. Process data: 1.\n",
    "Extract 3-mers: 1. Split train/test dataset with split ratio 0.7 1.\n",
    "Extract features : We used Latent Dirichlet Allocation to extract\n",
    "patterns from genome data. 1. Classify: We used Random Forest Classifier\n",
    "to find a mapping from extracted topic distributions to labels\n",
    "(multiclass problem). 1. Evaluation: We use (accuracy/f1) measure as our\n",
    "evaluation"
   ]
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
