{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Topic Modeling with SARS-Cov-2 Genome ðŸ§¬\n",
    "=======================================\n",
    "\n",
    "Video link: https://kth.box.com/s/y3jsb9lgp6cll6op15o6z77rchaefh24\n",
    "\n",
    "#### Problem description:\n",
    "\n",
    "SARS-CoV-2 is spreading across the world and as it spreads mutations are\n",
    "occuring. A way to understand the spreading and the mutations is to\n",
    "explore the structure and information hidden in the genome.\n",
    "\n",
    "#### Project goal:\n",
    "\n",
    "The Goal of this project is to explore a SARS-CoV-2 genome dataset and\n",
    "try to predict the origin of a SARS-CoV-2 genome sample.\n",
    "\n",
    "#### Data:\n",
    "\n",
    "We will use publicly available NCBI SARS-CoV-2 genome with their\n",
    "geographic region information.\n",
    "\n",
    "| Geographic region | \\#samples | |:-----------------: |:--------: | |\n",
    "Africa | 397 | | Asia | 2534 | | Europe | 1418 | | North America | 26836\n",
    "| | Oceania | 13304 | | South America | 158 |\n",
    "\n",
    "Data link:\n",
    "https://www.ncbi.nlm.nih.gov/labs/virus/vssi/\\#/virus?SeqType*s=Nucleotide&VirusLineage*ss=Severe%20acute%20respiratory%20syndrome%20coronavirus%202%20(SARS-CoV-2),%20taxid:2697049\n",
    "\n",
    "#### Background:\n",
    "\n",
    "-   Genome: Sequence of nucleotides (A-T-G-C)\n",
    "    \\[https://en.wikipedia.org/wiki/Genome\\]\n",
    "-   k-mer: Subsequences of length k of a genome\n",
    "    \\[https://en.wikipedia.org/wiki/K-mer\\]\n",
    "-   Sequence analysis of SARS-CoV-2 genome reveals features important\n",
    "    for vaccine design\n",
    "    \\[https://www.nature.com/articles/s41598-020-72533-2\\]\n",
    "-   Latent Dirichlet Allocation (LDA) tutorial from the course\n",
    "\n",
    "#### Challenges:\n",
    "\n",
    "-   How to encode genome sequence?\n",
    "    -   *Project solution :*\n",
    "        -   Represent genome as k-mers and use countVectorizer to\n",
    "            convert k-mers into a matrix of token counts (term-frequency\n",
    "            table)\n",
    "        -   Extract features with Latent Dirichlet Allocation (LDA) by\n",
    "            considering each genome sequence as a document and each\n",
    "            3-mer as a word. So, we have a collection of genomes\n",
    "            consisting of 3-mers.\n",
    "-   How to relate encoded features to the origins?\n",
    "    -   *Project solution:* We used a Random Forest Classifier and tried\n",
    "        both topic distributions, LDA output, and k-mer frequencies\n",
    "        directly. One advantage is interpretability: we can understand\n",
    "        the positive or negative relations a topic has on the origin.\n",
    "-   How to solve unbalanced class problem? E.g. North America has 26836\n",
    "    samples but South America has only 158\n",
    "    -   *Project solution:* Use f1 measure as metric\n",
    "\n",
    "#### Project steps:\n",
    "\n",
    "1.  Get SARS-CoV-2 data from NCBI\n",
    "2.  Process data:\n",
    "3.  Extract 3-mers:\n",
    "4.  Split train/test dataset with split ratio 0.7\n",
    "5.  Extract topic features: We used Latent Dirichlet Allocation to\n",
    "    extract patterns from k-mers features.\n",
    "6.  Classify: We used Random Forest Classifier\n",
    "    -   (Classification directly on k-mers features) To find a mapping\n",
    "        from extracted k-mers features to labels (multiclass problem).\n",
    "    -   (Classification on LDA features) To find a mapping from\n",
    "        extracted topic distributions to labels (multiclass problem).\n",
    "7.  Evaluation: We use (accuracy and f1) measure as our evaluation\n",
    "    metrics. We compared Classification on LDA features vs\n",
    "    Classification directly on k-mers features to see whether LDA is\n",
    "    capable of summarizing the data (and thus reducing the feature\n",
    "    dimensionality)\n",
    "\n",
    "#### What we lack mainly:\n",
    "\n",
    "-   A biological interpretation of the results (whether found terms in\n",
    "    topic distributions are significant/connected in a biological\n",
    "    network)."
   ]
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
