
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Genomics Analysis with Glow and Spark &#8212; Scalable Data Science &amp; Distributed Machine Learning</title>
    
  <link rel="stylesheet" href="../../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../../index.html">
  
  <img src="../../_static/logo.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">Scalable Data Science & Distributed Machine Learning</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/000_ScaDaMaLe.html">
   Introduction
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Introduction to Apache Spark Core
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/001_whySpark.html">
   Why Apache Spark?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html">
   databricks community edition
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#essentials-of-databricks-cloud-dbc-in-a-big-hurry">
   Essentials of Databricks Cloud (DBC) in a Big Hurry
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#dbc-essentials-what-is-databricks-cloud">
   DBC Essentials: What is Databricks Cloud?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#dbc-essentials-shard-cluster-notebook-and-dashboard">
   DBC Essentials: Shard, Cluster, Notebook and Dashboard
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#dbc-essentials-team-state-collaboration-elastic-resources">
   DBC Essentials: Team, State, Collaboration, Elastic Resources
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#you-should-all-have-databricks-community-edition-account-by-now">
   You Should All Have databricks community edition account by now!
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#import-course-content-now">
   Import Course Content Now!
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_00_loginToDatabricks.html#cloud-free-computing-environment-optional-but-recommended">
   Cloud-free Computing Environment (Optional but recommended)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_01_multiLingualNotebooks.html">
   Notebooks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/002_01_multiLingualNotebooks.html#further-reference-homework-recurrrent-points-of-reference">
   Further Reference / Homework / Recurrrent Points of Reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html">
   Introduction to Scala through Scala Notebook
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#scala-in-your-own-computer">
   Scala in your own computer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#scala-resources">
   Scala Resources
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#introduction-to-scala">
   Introduction to Scala
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#let-s-get-our-hands-dirty-in-scala">
   Let’s get our hands dirty in Scala
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#scala-types">
   Scala Types
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#expressions">
   Expressions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#blocks">
   Blocks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#functions">
   Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#methods">
   Methods
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#classes">
   Classes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#case-classes">
   Case Classes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#objects">
   Objects
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#traits">
   Traits
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#main-method">
   Main Method
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_00_scalaCrashCourse.html#what-i-try-not-do-while-learning-a-new-language">
   What I try not do while learning a new language?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html">
   Scala Resources
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html#let-s-continue-to-get-our-hands-dirty-in-scala">
   Let’s continue to get our hands dirty in Scala
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html#scala-type-hierarchy">
   Scala Type Hierarchy
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html#scala-collections">
   Scala Collections
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html#exercise-in-functional-programming">
   Exercise in Functional Programming
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html#lazy-evaluation">
   Lazy Evaluation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/003_01_scalaCrashCourse.html#recursions">
   Recursions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/004_RDDsTransformationsActions.html">
   Introduction to Spark
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/004_RDDsTransformationsActions.html#spark-cluster-overview">
   Spark Cluster Overview:
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/005_RDDsTransformationsActionsHOMEWORK.html">
   HOMEWORK notebook - RDDs Transformations and Actions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html">
   Piped RDDs and Bayesian AB Testing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#from-a-local-collection">
   From a Local Collection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#glom">
   glom
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#checkpointing">
   Checkpointing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#pipe-rdds-to-system-commands">
   Pipe RDDs to System Commands
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#mappartitions">
   mapPartitions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#foreachpartition">
   foreachPartition
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#numerically-rigorous-bayesian-ab-testing">
   Numerically Rigorous Bayesian AB Testing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006a_PipedRDD.html#usage-instructions-for-isit1or2coins">
   Usage instructions for IsIt1or2Coins
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/006_WordCount.html">
   Word Count on US State of the Union (SoU) Addresses
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Introduction to Apache Spark SQL
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007_SparkSQLIntroBasics.html">
   Introduction to Spark SQL
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007_SparkSQLIntroBasics.html#overview">
   Overview
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007_SparkSQLIntroBasics.html#datasets-and-dataframes">
   Datasets and DataFrames
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007_SparkSQLIntroBasics.html#getting-started-in-spark-2-x">
   Getting Started in Spark 2.x
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007a_SparkSQLProgGuide_HW.html">
   Spark Sql Programming Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007b_SparkSQLProgGuide_HW.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007b_SparkSQLProgGuide_HW.html#id1">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007c_SparkSQLProgGuide_HW.html">
   Getting Started - Exercise
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007d_SparkSQLProgGuide_HW.html">
   Data Sources
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007d_SparkSQLProgGuide_HW.html#id1">
   Data Sources
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007e_SparkSQLProgGuide_HW.html">
   Performance Tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007f_SparkSQLProgGuide_HW.html">
   Distributed SQL Engine
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007f_SparkSQLProgGuide_HW.html#id1">
   Distributed SQL Engine
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html">
   ScaDaMaLe, Scalable Data Science and Distributed Machine Learning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html#sql-pivoting-since-spark-2-4">
   SQL Pivoting since Spark 2.4
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html#load-data">
   Load Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html#pivoting-in-sql">
   Pivoting in SQL
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html#pivoting-with-multiple-aggregate-expressions">
   Pivoting with Multiple Aggregate Expressions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html#pivoting-with-multiple-grouping-columns">
   Pivoting with Multiple Grouping Columns
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/007g_PivotInSQL.html#pivoting-with-multiple-pivot-columns">
   Pivoting with Multiple Pivot Columns
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/008_DiamondsPipeline_01ETLEDA.html">
   Diamonds ML Pipeline Workflow - DataFrame ETL and EDA Part
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/008_DiamondsPipeline_01ETLEDA.html#step-1-load-data-as-dataframe">
   Step 1. Load data as DataFrame
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/008_DiamondsPipeline_01ETLEDA.html#step-2-understand-the-data">
   Step 2. Understand the data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/008_DiamondsPipeline_01ETLEDA.html#let-us-run-through-some-basic-inteactive-sql-queries-next">
   Let us run through some basic inteactive SQL queries next
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/008_DiamondsPipeline_01ETLEDA.html#why-do-we-need-to-know-these-interactive-sql-queries">
   Why do we need to know these interactive SQL queries?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/008_DiamondsPipeline_01ETLEDA.html#we-will-continue-later-with-ml-pipelines-to-do-prediction-with-a-fitted-model-from-this-dataset">
   We will continue later with ML pipelines to do prediction with a fitted model from this dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/009_PowerPlantPipeline_01ETLEDA.html">
   Power Plant ML Pipeline Application - DataFrame Part
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/009_PowerPlantPipeline_01ETLEDA.html#step-1-business-understanding">
   Step 1: Business Understanding
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/009_PowerPlantPipeline_01ETLEDA.html#step-2-load-your-data">
   Step 2: Load Your Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/009_PowerPlantPipeline_01ETLEDA.html#use-this-for-other-smallish-datasets-you-want-to-import-to-your-ce">
   USE THIS FOR OTHER SMALLish DataSets you want to import to your CE
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/009_PowerPlantPipeline_01ETLEDA.html#step-3-explore-your-data">
   Step 3: Explore Your Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/009_PowerPlantPipeline_01ETLEDA.html#step-4-visualize-your-data">
   Step 4: Visualize Your Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/010_wikipediaClickStream_01ETLEDA.html">
   Wiki Clickstream Analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/010_wikipediaClickStream_01ETLEDA.html#wikipedia-logo">
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_LoadExtract.html">
   Old Bailey Online Data Analysis in Apache Spark
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_LoadExtract.html#analysing-the-full-old-bailey-online-sessions-papers-dataset">
   Analysing the Full Old Bailey Online Sessions Papers Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html">
   Piped RDDs and Bayesian AB Testing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#from-a-local-collection">
   From a Local Collection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#glom">
   glom
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#checkpointing">
   Checkpointing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#pipe-rdds-to-system-commands">
   Pipe RDDs to System Commands
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#mappartitions">
   mapPartitions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#foreachpartition">
   foreachPartition
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#numerically-rigorous-bayesian-ab-testing">
   Numerically Rigorous Bayesian AB Testing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#usage-instructions-for-isit1or2coins">
   Usage instructions for IsIt1or2Coins
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#parsing-the-output-from-isit1or2coins">
   Parsing the output from
   <code class="docutils literal notranslate">
    <span class="pre">
     IsIt1or2Coins
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/033_OBO_PipedRDD_RigorousBayesianABTesting.html#providing-case-classes-for-input-and-output-for-easy-spark-communication">
   Providing case classes for input and output for easy spark communication
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/035_LDA_CornellMovieDialogs.html">
   Topic Modeling of Movie Dialogs with Latent Dirichlet Allocation
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Distribute Deep Learning with Horovod
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_7-sds-3-x-ddl/00_DDL_Introduction.html">
   Introduction to Distributed Deep Learning (DDL) with Horovod over Tensorflow/keras and Pytorch
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_7-sds-3-x-ddl/0x_mnist-tensorflow-keras.html">
   Distributed deep learning training using TensorFlow and Keras with HorovodRunner
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_7-sds-3-x-ddl/0y_mnist-pytorch.html">
   Distributed deep learning training using PyTorch with HorovodRunner for MNIST
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  WASP AI-Track Student Projects
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/035_LDA_CornellMovieDialogs.html">
   Topic Modeling of Movie Dialogs with Latent Dirichlet Allocation
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Voluntary Student Projects
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/030_Spark_GDELT_project.html">
   Plugging into GDELT Streams - TODO - IN PROGRESS
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/030_Spark_GDELT_project.html#this-is-just-dipping-our-pinky-toe-in-this-ocean-of-information">
   This is just dipping our pinky toe in this ocean of information!
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../000_1-sds-3-x/030_Spark_GDELT_project.html#download-from-gdelt-project">
   Download from gdelt-project
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/000_0-sds-3-x-projects/student-project-13_group-Genomics/02_1000genomes.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://lamastex.github.io/ScaDaMaLe/index.html"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://lamastex.github.io/ScaDaMaLe/index.html/issues/new?title=Issue%20on%20page%20%2F000_0-sds-3-x-projects/student-project-13_group-Genomics/02_1000genomes.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Genomics Analysis with Glow and Spark
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#problem-background">
     Problem background
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data">
     Data
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#method">
     Method
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-libs-and-define-helper-functions">
   Load libs and define helper functions
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#setup-and-loading-data">
     Setup and loading data
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#read-data">
   Read data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-exploration-and-filtering">
     Data exploration and filtering
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pca">
   PCA
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#read-sample-metadata-and-add-to-pca-components">
     Read sample metadata and add to PCA components
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#predicting-ethinicity">
   Predicting Ethinicity
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#filtering-of-snps-based-on-chi-squared-test">
   Filtering of SNPs based on chi-squared test
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#transformation-of-dataframe-to-fit-chisqselector">
     Transformation of dataframe to fit ChiSqSelector
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fitting-chisqselector">
     Fitting ChiSqSelector
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#train-logistic-regression-and-random-forest-models">
     Train logistic regression and random forest models
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#results-and-discussion">
     Results and Discussion
    </a>
   </li>
  </ul>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="genomics-analysis-with-glow-and-spark">
<h1>Genomics Analysis with Glow and Spark<a class="headerlink" href="#genomics-analysis-with-glow-and-spark" title="Permalink to this headline">¶</a></h1>
<p>The aim of this notebook is to analyze genomic data in the form of SNPs,
and see how different variations of SNPs correlated to ethnicity. This
work is inspired by the paper from Huang et al., <a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-2328-0">Genetic differences
among ethnic
groups</a>
(2015), and the notebook
https://glow.readthedocs.io/en/latest/_static/notebooks/tertiary/gwas.html.</p>
<div class="section" id="problem-background">
<h2>Problem background<a class="headerlink" href="#problem-background" title="Permalink to this headline">¶</a></h2>
<p>Each person as a unique setup of DNA. The DNA consitst of necleotides,
structured as a double helix, where each neucliotide binds to one other.
The DNA is split between 22 chromosomes, and each person has 23 pairs of
chorosomes. There are four different neucliotides, commonly denoted as
A, T, C, and G.</p>
<p>Single nucleotide polymorphisms (SNPs) are the most common genetic
variation between individuals. Each SNP represents a variation of a
specific neucleotide. For example, a SNP may replace the nucleotide
cytosine (C) with the nucleotide thymine (T) in a certain stretch of
DNA.</p>
</div>
<div class="section" id="data">
<h2>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h2>
<p>Genomic data is collected from the <a class="reference external" href="https://www.internationalgenome.org/">1000 Genomes
project</a>, with corresponding
sample annotations for all individuals in the dataset. For simiplicty,
we are only analyzing SNPs assosiated to chromosome 1, however this
study can easily be extended to include SNPs from all chromosomes.</p>
<p>The data consists of approximatly 6.5 million SNPs from 2504 subjects.</p>
</div>
<div class="section" id="method">
<h2>Method<a class="headerlink" href="#method" title="Permalink to this headline">¶</a></h2>
<p>After reading the data, we filter low quality SNPs. After this
operation, we end up with approx. 400’000 SNPs.</p>
<p>By doing a correlation analysis using PCA, we see that different
ethnicities cluster together. There is therfore a good reason to suppose
that SNPs can be used to predict ethnicity. However, since not all SNPs
are correlated to ethnicity, we want to only use the most relevant ones
for linear regression analysis.</p>
<p>For each SNPs, we calculate the correlation between the values and
ethnicity, and take the SNPs with a higher correlation than a threshold
value of 0.6 (or maximum 2000 SNPs).</p>
<p>Using the selected SNPs as features, we do a linear regression analysis.
We make some plots.</p>
</div>
</div>
<div class="section" id="load-libs-and-define-helper-functions">
<h1>Load libs and define helper functions<a class="headerlink" href="#load-libs-and-define-helper-functions" title="Permalink to this headline">¶</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">pyspark.sql.functions</span> <span class="kn">import</span> <span class="n">array_min</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">monotonically_increasing_id</span><span class="p">,</span> <span class="n">when</span><span class="p">,</span> <span class="n">log10</span>
<span class="kn">from</span> <span class="nn">pyspark.sql.types</span> <span class="kn">import</span> <span class="n">StringType</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.linalg</span> <span class="kn">import</span> <span class="n">Vector</span><span class="p">,</span> <span class="n">Vectors</span><span class="p">,</span> <span class="n">SparseVector</span><span class="p">,</span> <span class="n">DenseMatrix</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.stat</span> <span class="kn">import</span> <span class="n">Summarizer</span>
<span class="kn">from</span> <span class="nn">pyspark.mllib.linalg.distributed</span> <span class="kn">import</span> <span class="n">RowMatrix</span>
<span class="kn">from</span> <span class="nn">pyspark.mllib.util</span> <span class="kn">import</span> <span class="n">MLUtils</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.feature</span> <span class="kn">import</span> <span class="n">IndexToString</span><span class="p">,</span> <span class="n">StringIndexer</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.feature</span> <span class="kn">import</span> <span class="n">OneHotEncoder</span>
<span class="kn">from</span> <span class="nn">pyspark.sql.functions</span> <span class="kn">import</span> <span class="n">col</span><span class="p">,</span><span class="n">lit</span>

<span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>

<span class="kn">import</span> <span class="nn">mlflow</span>
<span class="kn">import</span> <span class="nn">glow</span>
<span class="n">glow</span><span class="o">.</span><span class="n">register</span><span class="p">(</span><span class="n">spark</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Helper functions</span>

<span class="k">def</span> <span class="nf">plot_layout</span><span class="p">(</span><span class="n">plot_title</span><span class="p">,</span> <span class="n">plot_style</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">):</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="n">plot_style</span><span class="p">)</span> <span class="c1">#e.g. ggplot, seaborn-colorblind, print(plt.style.available)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">plot_title</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;$</span><span class="si">{0}</span><span class="s1">$&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">xlabel</span><span class="p">))</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;bottom&#39;</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
  
<span class="k">def</span> <span class="nf">plot_histogram</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">,</span> <span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">nbins</span><span class="p">,</span> <span class="n">plot_title</span><span class="p">,</span> <span class="n">plot_style</span><span class="p">,</span> <span class="n">color</span><span class="p">,</span> <span class="n">vline</span><span class="p">,</span> <span class="n">out_path</span><span class="p">):</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
  <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">nbins</span><span class="p">)</span>
  <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">toPandas</span><span class="p">()</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">],</span> <span class="n">bins</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">vline</span><span class="p">:</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">vline</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
  <span class="n">plot_layout</span><span class="p">(</span><span class="n">plot_title</span><span class="p">,</span> <span class="n">plot_style</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">out_path</span><span class="p">)</span>
  <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
  
<span class="k">def</span> <span class="nf">calculate_pval_bonferroni_cutoff</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
  <span class="n">bonferroni_p</span> <span class="o">=</span>  <span class="n">cutoff</span> <span class="o">/</span> <span class="n">df</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
  <span class="k">return</span> <span class="n">bonferroni_p</span>

<span class="k">def</span> <span class="nf">get_sample_info</span><span class="p">(</span><span class="n">vcf_df</span><span class="p">,</span> <span class="n">sample_metadata_df</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  get sample IDs from VCF dataframe, index them, then join to sample metadata dataframe</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">sample_id_list</span> <span class="o">=</span> <span class="n">vcf_df</span><span class="o">.</span><span class="n">limit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;genotypes.sampleId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">(</span><span class="s2">&quot;sampleId&quot;</span><span class="p">)</span>
  <span class="n">sample_id_indexed</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">sample_id_list</span><span class="p">,</span> <span class="n">StringType</span><span class="p">())</span><span class="o">.</span> \
                            <span class="n">coalesce</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span> \
                            <span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;value&quot;</span><span class="p">,</span> <span class="s2">&quot;Sample&quot;</span><span class="p">)</span><span class="o">.</span> \
                            <span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="n">monotonically_increasing_id</span><span class="p">())</span>
  <span class="n">sample_id_annotated</span> <span class="o">=</span> <span class="n">sample_id_indexed</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_metadata_df</span><span class="p">,</span> <span class="s2">&quot;Sample&quot;</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">sample_id_annotated</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Paths to store/find data.</span>
<span class="c1"># Since a lot of the processing takes a long time, we store intermediate results.</span>
<span class="n">vcf_path</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz&quot;</span>
<span class="n">delta_silver_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/gwas_test/snps.delta&quot;</span>

<span class="n">gwas_results_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/gwas_test/gwas_results.delta&quot;</span>
<span class="n">phenotype_path</span> <span class="o">=</span> <span class="s2">&quot;/databricks-datasets/genomics/1000G/phenotypes.normalized&quot;</span>
<span class="n">sample_info_path</span> <span class="o">=</span> <span class="s2">&quot;/databricks-datasets/genomics/1000G/samples/populations_1000_genomes_samples.csv&quot;</span>

<span class="n">principal_components_path</span> <span class="o">=</span> <span class="s2">&quot;/dbfs/datasets/sds/genomics/pcs.delta&quot;</span>
<span class="n">hwe_path</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/hwe.delta&quot;</span>
<span class="n">vectorized_path</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/vectorized.delta&quot;</span>
<span class="n">delta_gold_path</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/snps.qced.delta.delta&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="setup-and-loading-data">
<h2>Setup and loading data<a class="headerlink" href="#setup-and-loading-data" title="Permalink to this headline">¶</a></h2>
<p>The data used was dowloaded from
ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr1.phase3<em>shapeit2</em>mvncall<em>integrated</em>v5a.20130502.genotypes.vcf.gz</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="mi">1000</span><span class="n">genomes</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">ftp</span><span class="o">/</span><span class="n">release</span><span class="o">/</span><span class="mi">20130502</span><span class="o">/</span><span class="n">ALL</span><span class="o">.</span><span class="n">chr1</span><span class="o">.</span><span class="n">phase3_shapeit2_mvncall_integrated_v5a</span><span class="o">.</span><span class="mf">20130502.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="read-data">
<h1>Read data<a class="headerlink" href="#read-data" title="Permalink to this headline">¶</a></h1>
<p>The data is read using the <a class="reference external" href="https://projectglow.io/">Glow</a>, an
open-source library for working with genomics data. It is inlcuded when
enabling “Databricks Runtime for Genomics”, allowing easy read of
genomic-specific file formats, and other helper methods.</p>
<div class="section" id="data-exploration-and-filtering">
<h2>Data exploration and filtering<a class="headerlink" href="#data-exploration-and-filtering" title="Permalink to this headline">¶</a></h2>
<p>Load and view the vcf files. The info fields are combined to one column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vcf_view_unsplit</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;vcf&quot;</span><span class="p">)</span><span class="o">.</span> \
   <span class="n">option</span><span class="p">(</span><span class="s2">&quot;flattenInfoFields&quot;</span><span class="p">,</span> <span class="s2">&quot;false&quot;</span><span class="p">)</span><span class="o">.</span> \
   <span class="n">load</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">vcf_view_unsplit</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;genotypes&quot;</span><span class="p">,</span> <span class="n">col</span><span class="p">(</span><span class="s2">&quot;genotypes&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>
</pre></div>
</div>
</div>
</div>
<p>In the dataframe above, we see that we have columns named
“referenceAllele” and “alternateAlleles”. The data so called variations,
i.e., genetic sequences which are different between two individuals. The
difference may appear differently, and each difference is called an
allele. In the data, we have reference genomes, and the alternate
alleles are all the variations at a specific position which is found
amoung the analyzed subjects.</p>
<p>We do not want multiple alternative allelels in one row, so we split
them using the <code class="docutils literal notranslate"><span class="pre">split_miltiallelics</span></code> function from Glow.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vcf_view</span> <span class="o">=</span> <span class="n">glow</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;split_multiallelics&quot;</span><span class="p">,</span> <span class="n">vcf_view_unsplit</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">vcf_view</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;genotypes&quot;</span><span class="p">,</span> <span class="n">col</span><span class="p">(</span><span class="s2">&quot;genotypes&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>
</pre></div>
</div>
</div>
</div>
<p>We now save our modified dataframe in the Delta format (which compared
to VCF is more user friendly). At the same time, we calulcate som
neccessary statistics, which we will use later, using the Glow functions
<code class="docutils literal notranslate"><span class="pre">call_summary_stats</span></code> and <code class="docutils literal notranslate"><span class="pre">hardy_weinberg</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># NOTE - this takes approx 2 hours</span>

<span class="c1"># vcf_view.selectExpr(&quot;*&quot;, &quot;expand_struct(call_summary_stats(genotypes))&quot;, &quot;expand_struct(hardy_weinberg(genotypes))&quot;). \</span>
<span class="c1">#    write. \</span>
<span class="c1">#    mode(&quot;overwrite&quot;). \</span>
<span class="c1">#    format(&quot;delta&quot;). \</span>
<span class="c1">#    save(delta_silver_path)</span>
</pre></div>
</div>
</div>
</div>
<p>The statistics we calculated, as well as the Hardy-Weinberg equilibrium
p-values (which basically denotes the probability of a given allele is
probable to be true or may be a reading mistake), are used to filter out
low quality SNPs.</p>
<p>We read the saved dataframe, and filter the dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Hyper paramters</span>
<span class="n">allele_freq_cutoff</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="n">num_pcs</span> <span class="o">=</span> <span class="mi">5</span> <span class="c1">#number of principal components</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># hwe = spark.read.format(&quot;delta&quot;). \</span>
<span class="c1">#                  load(delta_silver_path). \</span>
<span class="c1">#                  where((col(&quot;alleleFrequencies&quot;).getItem(0) &gt;= allele_freq_cutoff) &amp; </span>
<span class="c1">#                        (col(&quot;alleleFrequencies&quot;).getItem(0) &lt;= (1.0 - allele_freq_cutoff))). \</span>
<span class="c1">#                  withColumn(&quot;log10pValueHwe&quot;, when(col(&quot;pValueHwe&quot;) == 0, 26).otherwise(-log10(col(&quot;pValueHwe&quot;))))</span>

<span class="c1"># hwe.write. \</span>
<span class="c1">#    mode(&quot;overwrite&quot;). \</span>
<span class="c1">#    format(&quot;delta&quot;).save(hwe_path)</span>

<span class="n">hwe</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">hwe_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">hwe_cutoff</span> <span class="o">=</span> <span class="n">calculate_pval_bonferroni_cutoff</span><span class="p">(</span><span class="n">hwe</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">plot_histogram</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">hwe</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;log10pValueHwe&quot;</span><span class="p">),</span> 
                       <span class="n">col</span><span class="o">=</span><span class="s2">&quot;log10pValueHwe&quot;</span><span class="p">,</span>
                       <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;-log_</span><span class="si">{10}</span><span class="s1">(P)&#39;</span><span class="p">,</span>
                       <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> 
                       <span class="n">xmax</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> 
                       <span class="n">nbins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> 
                       <span class="n">plot_title</span><span class="o">=</span><span class="s2">&quot;hardy-weinberg equilibrium&quot;</span><span class="p">,</span> 
                       <span class="n">plot_style</span><span class="o">=</span><span class="s2">&quot;ggplot&quot;</span><span class="p">,</span>
                       <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#e41a1c&#39;</span><span class="p">,</span>
                       <span class="n">vline</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">hwe_cutoff</span><span class="p">),</span>
                       <span class="n">out_path</span> <span class="o">=</span> <span class="s2">&quot;/databricks/driver/hwe.png&quot;</span>
                      <span class="p">)</span>
       <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Filter and save new dataframe, only alleles with in the frequency band,
and where the Hardy-Weiburg value is higher than cutoff.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># spark.read.format(&quot;delta&quot;). \</span>
<span class="c1">#    load(hwe_path). \</span>
<span class="c1">#    where((col(&quot;alleleFrequencies&quot;).getItem(0) &gt;= allele_freq_cutoff) &amp; </span>
<span class="c1">#          (col(&quot;alleleFrequencies&quot;).getItem(0) &lt;= (1.0 - allele_freq_cutoff)) &amp;</span>
<span class="c1">#          (col(&quot;pValueHwe&quot;) &gt;= hwe_cutoff)). \</span>
<span class="c1">#    write. \</span>
<span class="c1">#    mode(&quot;overwrite&quot;). \</span>
<span class="c1">#    format(&quot;delta&quot;). \</span>
<span class="c1">#    save(delta_gold_path)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">hwe_filtered</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">delta_gold_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="pca">
<h1>PCA<a class="headerlink" href="#pca" title="Permalink to this headline">¶</a></h1>
<p>We perform a PCA analysis for data exploration purposes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># vectorized = spark.read.format(&quot;delta&quot;). \</span>
<span class="c1">#                         load(delta_gold_path). \</span>
<span class="c1">#                         selectExpr(&quot;array_to_sparse_vector(genotype_states(genotypes)) as features&quot;). \</span>
<span class="c1">#                         cache()</span>

<span class="c1"># vectorized.write. \</span>
<span class="c1">#    mode(&quot;overwrite&quot;). \</span>
<span class="c1">#    format(&quot;delta&quot;).save(&quot;dbfs:///datasets/sds/genomics/vectorized.delta&quot;)</span>

<span class="n">vectorized</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">vectorized_path</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">vectorized</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Note - takes approx 30 min</span>

<span class="c1"># matrix = RowMatrix(MLUtils.convertVectorColumnsFromML(vectorized, &quot;features&quot;).rdd.map(lambda x: x.features))</span>
<span class="c1"># pcs = matrix.computeSVD(num_pcs)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># @dataclass()</span>
<span class="c1"># class Covariates:</span>
<span class="c1">#     covariates: DenseMatrix</span>
      
<span class="c1"># spark.createDataFrame([Covariates(pcs.V.asML())]). \</span>
<span class="c1">#       write. \</span>
<span class="c1">#       format(&quot;delta&quot;). \</span>
<span class="c1">#       save(principal_components_path)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># pcs_df = spark.createDataFrame(pcs.V.toArray().tolist(), [&quot;pc&quot; + str(i) for i in range(num_pcs)])</span>

<span class="c1"># display(pcs_df)</span>

<span class="c1"># pcs_df.coalesce(1).write.format(&quot;com.databricks.spark.csv&quot;).option(&quot;header&quot;, &quot;true&quot;).save(&quot;dbfs:///datasets/sds/genomics/pcs_df.csv&quot;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read already caluclated pca</span>
<span class="n">pcs_df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;csv&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;dbfs:///datasets/sds/genomics/pcs_df.csv&quot;</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">pcs_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="read-sample-metadata-and-add-to-pca-components">
<h2>Read sample metadata and add to PCA components<a class="headerlink" href="#read-sample-metadata-and-add-to-pca-components" title="Permalink to this headline">¶</a></h2>
<p>We load the subject meta data (which includes information about
ethnicity).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s2">&quot;header&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="n">sample_info_path</span><span class="p">)</span>
<span class="n">sample_info</span> <span class="o">=</span> <span class="n">get_sample_info</span><span class="p">(</span><span class="n">vcf_view</span><span class="p">,</span> <span class="n">sample_metadata</span><span class="p">)</span>

<span class="n">sample_count</span> <span class="o">=</span> <span class="n">sample_info</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>

<span class="n">pcs_indexed</span> <span class="o">=</span> <span class="n">pcs_df</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="n">monotonically_increasing_id</span><span class="p">())</span>
<span class="n">pcs_with_samples</span> <span class="o">=</span> <span class="n">pcs_indexed</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_info</span><span class="p">,</span> <span class="s2">&quot;index&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>View 1st and 2nd principal component</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">pcs_with_samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We see that there there are some clustering based on ethnicity, showing
that it could be possible to tell ethnicity from the SNP information of
a subject.</p>
</div>
</div>
<div class="section" id="predicting-ethinicity">
<h1>Predicting Ethinicity<a class="headerlink" href="#predicting-ethinicity" title="Permalink to this headline">¶</a></h1>
<p>Replication of the paper “Genetic differences among ethnic groups”,
Huang et al.
https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-2328-0</p>
<p>Originally, we had approx. 6.5 milj genetic variations, from 2500
individuals. We first did a quality control by filtering only alleles
which occur frequently enough, and those above the Hardy-Weinberg P
value cutoff.</p>
<p>Since we still have over 400 000 varitations, we need to filter further.
As in referenced paper, we use the Cramér’s V statistic to measure the
correlation between ethnicity and genetic variation, with a value
between 0 and 1. Those with significantly high correlation are kept.</p>
<div class="math notranslate nohighlight">
\[\begin{split}V=\\sqrt{\\frac{\\chi^2/N}{ \\min \\left(k-1,r-1\\right)}}\end{split}\]</div>
<p>N is the number of individual samples, k is the number of ethic groups,
r is the SNP status (0, 1 or 2 minor alleles).</p>
<p>Where the Pearson’s chi-squared statistic can be calculated as:</p>
<div class="math notranslate nohighlight">
\[\begin{split}{\\chi}^2={\\displaystyle \\sum*{i=1}^k{\\displaystyle
\\sum*{j=1}^r\\frac{{\\left({O}*{i,j}-{E}*{i,j}\\right)}^2}{E\_{i,j}}}}\end{split}\]</div>
<p>Read the dataset that contains the population information and encode it
for regression</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read sample information</span>
<span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s2">&quot;header&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="n">sample_info_path</span><span class="p">)</span>
<span class="n">sample_info</span> <span class="o">=</span> <span class="n">get_sample_info</span><span class="p">(</span><span class="n">vcf_view</span><span class="p">,</span> <span class="n">sample_metadata</span><span class="p">)</span>

<span class="n">sample_count</span> <span class="o">=</span> <span class="n">sample_info</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>

<span class="n">mlflow</span><span class="o">.</span><span class="n">log_param</span><span class="p">(</span><span class="s2">&quot;number of samples&quot;</span><span class="p">,</span> <span class="n">sample_count</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">sample_info</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># One hot encoding of the labels</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.feature</span> <span class="kn">import</span> <span class="n">IndexToString</span><span class="p">,</span> <span class="n">StringIndexer</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.feature</span> <span class="kn">import</span> <span class="n">OneHotEncoder</span>

<span class="c1"># indexer = StringIndexer(inputCol=&quot;Population&quot;, outputCol=&quot;Population_index&quot;)</span>
<span class="n">indexer</span> <span class="o">=</span> <span class="n">StringIndexer</span><span class="p">(</span><span class="n">inputCol</span><span class="o">=</span><span class="s2">&quot;super_population&quot;</span><span class="p">,</span> <span class="n">outputCol</span><span class="o">=</span><span class="s2">&quot;Population_index&quot;</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">indexer</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">sample_info</span><span class="p">)</span>
<span class="n">indexed</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">sample_info</span><span class="p">)</span>

<span class="c1"># indexed.select(&quot;Population&quot;, &quot;Population_index&quot;).distinct().show(30)</span>

<span class="n">encoder</span> <span class="o">=</span> <span class="n">OneHotEncoder</span><span class="p">(</span><span class="n">inputCols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Population_index&quot;</span><span class="p">],</span>
                        <span class="n">outputCols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;population_onehot&quot;</span><span class="p">])</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">encoder</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">indexed</span><span class="p">)</span>
<span class="n">encoded</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">indexed</span><span class="p">)</span>
<span class="n">encoded</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="filtering-of-snps-based-on-chi-squared-test">
<h1>Filtering of SNPs based on chi-squared test<a class="headerlink" href="#filtering-of-snps-based-on-chi-squared-test" title="Permalink to this headline">¶</a></h1>
<p>According to <a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-2328-0">Tao Huang et al.
(2015)</a>,
85 % of SNPs are the same in all human populations, hence we will apply
chi-squared based feature selection to try to identify the approximately
15 % of SNPs that are population-specific. We decided to test our
classifiers with several sizes of the feature vectors: * all 416,005
available SNPs from Chromosome 1 * most relevant 200,000 SNPs * most
relevant 20,000 SNPs * most relevant 2,000 SNPs * most relevant 1,000
SNPs * most relevant 100 SNPs * most relevant 50 SNPs</p>
<p>This would give us a better understanding if the ChiSqSelector is
appropriate method for selecting the features in genomics. We aklowedge
that <a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-2328-0">Tao Huang et al.
(2015)</a>
used different metric based on chi-squared distribution, but
ChiSqSelector is the closest already implemented method in spark that we
could find.</p>
<p>In order to use ChiSqSelector from pyspark.ml.feature, we first need to
format the data into a sparce feature vectors and numeric corresponding
label.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load earlier-filtered data</span>
<span class="n">delta_gold_path</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/snps.qced.delta.delta&quot;</span>
<span class="n">hwe_filtered</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">delta_gold_path</span><span class="p">)</span>
<span class="n">vectorized_2</span> <span class="o">=</span> <span class="n">hwe_filtered</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">glow</span><span class="o">.</span><span class="n">genotype_states</span><span class="p">(</span><span class="s1">&#39;genotypes&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s1">&#39;states&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
<span class="n">vectorized_df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">vectorized_2</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">vectorized_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="transformation-of-dataframe-to-fit-chisqselector">
<h2>Transformation of dataframe to fit ChiSqSelector<a class="headerlink" href="#transformation-of-dataframe-to-fit-chisqselector" title="Permalink to this headline">¶</a></h2>
<p>We use monotonically<em>increasing</em>id, poseexplode and collect_list
methods to achieve the required format of the data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add a column that indicates to which SNP the states belong and then explode SNPs</span>
<span class="kn">from</span> <span class="nn">pyspark.sql.functions</span> <span class="kn">import</span> <span class="n">monotonically_increasing_id</span> 
<span class="kn">from</span> <span class="nn">pyspark.sql.functions</span> <span class="kn">import</span> <span class="n">explode</span><span class="p">,</span> <span class="n">posexplode</span>
<span class="kn">from</span> <span class="nn">pyspark.sql.functions</span> <span class="kn">import</span> <span class="n">col</span><span class="p">,</span> <span class="n">concat</span><span class="p">,</span> <span class="n">desc</span><span class="p">,</span> <span class="n">first</span><span class="p">,</span> <span class="n">lit</span><span class="p">,</span> <span class="n">row_number</span><span class="p">,</span> <span class="n">collect_list</span>

<span class="n">vec_df_dummy</span> <span class="o">=</span> <span class="n">vectorized_df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="n">monotonically_increasing_id</span><span class="p">())</span>
<span class="c1">#vec_exploded_states = vec_df_dummy.withColumn(&quot;expanded_states&quot;, explode(&quot;states&quot;))</span>
<span class="n">vec_exploded_states</span> <span class="o">=</span> <span class="n">vec_df_dummy</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="n">posexplode</span><span class="p">(</span><span class="s2">&quot;states&quot;</span><span class="p">))</span>
<span class="n">vec_exploded_states</span> <span class="o">=</span> <span class="n">vec_exploded_states</span><span class="o">.</span><span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;subjectID&quot;</span><span class="p">)</span>
<span class="n">vec_exploded_states</span> <span class="o">=</span> <span class="n">vec_exploded_states</span><span class="o">.</span><span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;col&quot;</span><span class="p">,</span> <span class="s2">&quot;expandedState&quot;</span><span class="p">)</span>
<span class="n">features_df</span> <span class="o">=</span> <span class="n">vec_exploded_states</span><span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;subjectID&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">collect_list</span><span class="p">(</span><span class="s2">&quot;expandedState&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;Features&quot;</span><span class="p">))</span>
<span class="n">features_df</span> <span class="o">=</span> <span class="n">features_df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">encoded</span><span class="p">,</span> <span class="n">features_df</span><span class="o">.</span><span class="n">subjectID</span> <span class="o">==</span> <span class="n">encoded</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;Features&quot;</span><span class="p">,</span><span class="s2">&quot;Population_index&quot;</span><span class="p">,</span> <span class="s2">&quot;population_onehot&quot;</span><span class="p">)</span>
<span class="n">features_df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, Glow utility function array<em>to</em>sparse_vector is used to
convert the dense feature vectors into sparse vectors</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">features_df</span> <span class="o">=</span> <span class="n">features_df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span><span class="s2">&quot;array_to_sparse_vector(Features) as features_sparse&quot;</span><span class="p">,</span><span class="s2">&quot;Population_index&quot;</span><span class="p">,</span> <span class="s2">&quot;population_onehot&quot;</span><span class="p">)</span>
<span class="n">features_df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="fitting-chisqselector">
<h2>Fitting ChiSqSelector<a class="headerlink" href="#fitting-chisqselector" title="Permalink to this headline">¶</a></h2>
<p>As the computation time of ChiSqSelector takes roughly 3 hours for each
subset of features, we are saving them to disc.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_selection_200_000</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/selected_feat_200_000.delta&quot;</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.feature</span> <span class="kn">import</span> <span class="n">ChiSqSelector</span>
<span class="c1"># selector = ChiSqSelector(featuresCol=&#39;features_sparse&#39;, outputCol=&#39;ChiSq&#39;,labelCol=&#39;Population_index&#39;, numTopFeatures = 200000)</span>
<span class="c1"># selected_feat_200_000 = selector.fit(features_df).transform(features_df)</span>
<span class="c1"># selected_feat_200_000.write.format(&quot;delta&quot;).save(feature_selection_200_000)</span>
<span class="n">selected_feat_200_000</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">feature_selection_200_000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_selection_20_000</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/selected_feat_20_000.delta&quot;</span>
<span class="c1"># selector = ChiSqSelector(featuresCol=&#39;features_sparse&#39;, outputCol=&#39;ChiSq&#39;,labelCol=&#39;Population_index&#39;, numTopFeatures = 20000)</span>
<span class="c1"># selected_feat_20_000 = selector.fit(features_df).transform(features_df)</span>
<span class="c1"># selected_feat_20_000.write.format(&quot;delta&quot;).save(feature_selection_20_000)</span>
<span class="n">selected_feat_20_000</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">feature_selection_20_000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_selection_2000</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/selected_feat_2000.delta&quot;</span>
<span class="c1"># from pyspark.ml.feature import ChiSqSelector</span>
<span class="c1"># selector = ChiSqSelector(featuresCol=&#39;features_sparse&#39;, outputCol=&#39;ChiSq&#39;,labelCol=&#39;Population_index&#39;, numTopFeatures = 2000)</span>
<span class="c1"># selected_feat_2000 = selector.fit(features_df).transform(features_df)</span>
<span class="c1"># selected_feat_2000.write.format(&quot;delta&quot;).save(feature_selection_2000)</span>

<span class="n">selected_feat_2000</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">feature_selection_2000</span><span class="p">)</span>
<span class="n">selected_feat_2000</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_selection_results_1000</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/selected_feat_1000.delta&quot;</span>
<span class="c1"># selector = ChiSqSelector(featuresCol=&#39;features_sparse&#39;, outputCol=&#39;ChiSq&#39;,labelCol=&#39;Population_index&#39;, numTopFeatures = 1000)</span>
<span class="c1"># selected_feat_1000 = selector.fit(features_df).transform(features_df)</span>
<span class="c1"># selected_feat_1000.write.format(&quot;delta&quot;).save(feature_selection_results_1000)</span>
<span class="c1">#result = selector.fit(features_df).transform(features_df)</span>
<span class="n">selected_feat_1000</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">feature_selection_results_1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_selection_results_100</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/selected_feat_100.delta&quot;</span>
<span class="c1"># selector = ChiSqSelector(featuresCol=&#39;features_sparse&#39;, outputCol=&#39;ChiSq&#39;,labelCol=&#39;Population_index&#39;, numTopFeatures = 100)</span>
<span class="c1"># selected_feat_100 = selector.fit(features_df).transform(features_df)</span>
<span class="c1"># selected_feat_100.write.format(&quot;delta&quot;).save(feature_selection_results_100)</span>
<span class="c1">#result = selector.fit(features_df).transform(features_df)</span>
<span class="n">selected_feat_100</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">feature_selection_results_100</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_selection_50</span> <span class="o">=</span> <span class="s2">&quot;dbfs:///datasets/sds/genomics/selected_feat_50.delta&quot;</span>
<span class="c1"># selector = ChiSqSelector(featuresCol=&#39;features_sparse&#39;, outputCol=&#39;ChiSq&#39;,labelCol=&#39;Population_index&#39;, numTopFeatures = 50)</span>
<span class="c1"># selected_feat_50 = selector.fit(features_df).transform(features_df)</span>
<span class="c1"># selected_feat_50.write.format(&quot;delta&quot;).save(feature_selection_50)</span>
<span class="n">selected_feat_50</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;delta&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">feature_selection_50</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="train-logistic-regression-and-random-forest-models">
<h2>Train logistic regression and random forest models<a class="headerlink" href="#train-logistic-regression-and-random-forest-models" title="Permalink to this headline">¶</a></h2>
<p>The code below implements a loop over datasets with different number of
SNPs, test-train split and fitting of logistic regression and random
forest models. The performance is measured in accuracy.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pyspark.ml.classification</span> <span class="kn">import</span> <span class="n">LogisticRegression</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.evaluation</span> <span class="kn">import</span> <span class="n">MulticlassClassificationEvaluator</span>
<span class="kn">from</span> <span class="nn">pyspark.ml.classification</span> <span class="kn">import</span> <span class="n">RandomForestClassifier</span>

<span class="k">def</span> <span class="nf">fit_ML_model</span><span class="p">(</span><span class="n">trainSet</span><span class="p">,</span><span class="n">testSet</span><span class="p">,</span> <span class="n">model_in</span><span class="p">):</span>

  <span class="c1"># Train model.  </span>
  <span class="n">model</span> <span class="o">=</span> <span class="n">model_in</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">trainSet</span><span class="p">)</span>

  <span class="c1"># Make predictions.</span>
  <span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">testSet</span><span class="p">)</span>

  <span class="c1"># Evaluate the classifier based on accuracy</span>
  <span class="n">evaluator</span> <span class="o">=</span> <span class="n">MulticlassClassificationEvaluator</span><span class="p">(</span>
      <span class="n">labelCol</span><span class="o">=</span><span class="s2">&quot;Population_index&quot;</span><span class="p">,</span> <span class="n">predictionCol</span><span class="o">=</span><span class="s2">&quot;prediction&quot;</span><span class="p">,</span> <span class="n">metricName</span><span class="o">=</span><span class="s2">&quot;accuracy&quot;</span><span class="p">)</span>
  <span class="n">accuracy</span> <span class="o">=</span> <span class="n">evaluator</span><span class="o">.</span><span class="n">evaluate</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">accuracy</span>


<span class="n">rf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">(</span><span class="n">labelCol</span><span class="o">=</span><span class="s2">&quot;Population_index&quot;</span><span class="p">,</span> <span class="n">featuresCol</span><span class="o">=</span><span class="s2">&quot;final_features&quot;</span><span class="p">,</span> <span class="n">numTrees</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">lr</span> <span class="o">=</span> <span class="n">LogisticRegression</span><span class="p">(</span><span class="n">featuresCol</span><span class="o">=</span><span class="s2">&quot;final_features&quot;</span><span class="p">,</span> <span class="n">labelCol</span><span class="o">=</span><span class="s2">&quot;Population_index&quot;</span><span class="p">,</span> <span class="n">maxIter</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">acc_rf</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">acc_lr</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Run the models on full features set (without ChiSqSelector)</span>
<span class="n">features_ready</span> <span class="o">=</span> <span class="n">selected_feat_2000</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span><span class="s2">&quot;features_sparse as final_features&quot;</span><span class="p">,</span><span class="s2">&quot;Population_index&quot;</span><span class="p">)</span>
<span class="n">trainSet</span><span class="p">,</span> <span class="n">testSet</span> <span class="o">=</span> <span class="n">features_ready</span><span class="o">.</span><span class="n">randomSplit</span><span class="p">((</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">),</span> <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
<span class="n">acc_rf</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fit_ML_model</span><span class="p">(</span><span class="n">trainSet</span><span class="p">,</span> <span class="n">testSet</span><span class="p">,</span> <span class="n">rf</span><span class="p">))</span>
<span class="n">acc_lr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fit_ML_model</span><span class="p">(</span><span class="n">trainSet</span><span class="p">,</span> <span class="n">testSet</span><span class="p">,</span> <span class="n">lr</span><span class="p">))</span>

<span class="c1"># Run the models on selected features by ChiSqSelector</span>
<span class="k">for</span> <span class="n">data_item</span> <span class="ow">in</span> <span class="p">[</span><span class="n">selected_feat_200_000</span><span class="p">,</span> <span class="n">selected_feat_20_000</span><span class="p">,</span> <span class="n">selected_feat_2000</span><span class="p">,</span> <span class="n">selected_feat_1000</span><span class="p">,</span> <span class="n">selected_feat_100</span><span class="p">,</span> <span class="n">selected_feat_50</span><span class="p">]:</span>
    <span class="c1"># Work around for the bug in ChiSqSelector - otherwise does not work with the random forest model</span>
    <span class="c1">#ChiSqSelector has a bug that it formats data in a way that RandomForest does not accept. We found this work around to work. Bug reported in https://stackoverflow.com/questions/46269275/spark-ml-issue-in-training-after-using-chisqselector-for-feature-selection</span>
    <span class="n">features_ready</span> <span class="o">=</span> <span class="n">data_item</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">glow</span><span class="o">.</span><span class="n">vector_to_array</span><span class="p">(</span><span class="s1">&#39;ChiSq&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s1">&#39;features_dense&#39;</span><span class="p">),</span> <span class="s2">&quot;Population_index&quot;</span><span class="p">,</span> <span class="s2">&quot;ChiSq&quot;</span><span class="p">)</span>
    <span class="n">features_ready</span> <span class="o">=</span> <span class="n">features_ready</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span><span class="s2">&quot;array_to_sparse_vector(features_dense) as final_features&quot;</span><span class="p">,</span><span class="s2">&quot;Population_index&quot;</span><span class="p">)</span>
    <span class="n">trainSet</span><span class="p">,</span> <span class="n">testSet</span> <span class="o">=</span> <span class="n">features_ready</span><span class="o">.</span><span class="n">randomSplit</span><span class="p">((</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">),</span> <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
    <span class="n">acc_rf</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fit_ML_model</span><span class="p">(</span><span class="n">trainSet</span><span class="p">,</span> <span class="n">testSet</span><span class="p">,</span> <span class="n">rf</span><span class="p">))</span>
    <span class="n">acc_lr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fit_ML_model</span><span class="p">(</span><span class="n">trainSet</span><span class="p">,</span> <span class="n">testSet</span><span class="p">,</span> <span class="n">lr</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="results-and-discussion">
<h2>Results and Discussion<a class="headerlink" href="#results-and-discussion" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">rf_plot</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;all_feat&#39;</span><span class="p">,</span> <span class="s1">&#39;200,000&#39;</span><span class="p">,</span> <span class="s1">&#39;20,000&#39;</span><span class="p">,</span> <span class="s1">&#39;2,000&#39;</span><span class="p">,</span> <span class="s1">&#39;1,000&#39;</span><span class="p">,</span> <span class="s1">&#39;100&#39;</span><span class="p">,</span> <span class="s1">&#39;50&#39;</span><span class="p">],</span> <span class="n">y</span><span class="o">=</span><span class="n">acc_rf</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="n">lf_plot</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;all_feat&#39;</span><span class="p">,</span> <span class="s1">&#39;200,000&#39;</span><span class="p">,</span> <span class="s1">&#39;20,000&#39;</span><span class="p">,</span> <span class="s1">&#39;2,000&#39;</span><span class="p">,</span> <span class="s1">&#39;1,000&#39;</span><span class="p">,</span> <span class="s1">&#39;100&#39;</span><span class="p">,</span> <span class="s1">&#39;50&#39;</span><span class="p">],</span> <span class="n">y</span><span class="o">=</span><span class="n">acc_lr</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Number of features&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Accuracy&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">((</span><span class="n">rf_plot</span><span class="p">,</span> <span class="n">lf_plot</span><span class="p">),</span>
           <span class="p">(</span><span class="s1">&#39;Random forest&#39;</span><span class="p">,</span> <span class="s1">&#39;Logistic regression&#39;</span><span class="p">),</span>
           <span class="n">scatterpoints</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
           <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
           <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
           <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>The plot above shows the accuracy of random forest and logistic
regression classifiers on predicting the ethnicity from prepared SNPs.
Both classifiers preformed much better than random guessing, hence the
ethnicity information is clearly encoded in SNPs. The logistic
regression consistently outperformed random forest and reached accuracy
over 90 %. We can also see that random forest was sensitive to the
feature selection and it’s performance droped once fewer that the all
available SNPs (416,005 SNPs) were used. In contrast, logistic
regression performed equally well with half of the available features
(200,000 SNPs). This indicates that having well-tuned classifier and an
appropriate feature selector allows us to reduce the required number of
features dramatically without compromising the performance. The work
could be improved by testing other ways of selecting the relevant SNPs,
tuning the hyperparameters in a grid-search manner, building confidence
intervals based on bootstrapping or cross-validation and testing other
types of classifiers.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./000_0-sds-3-x-projects/student-project-13_group-Genomics"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By ScaDaMaLe Team<br/>
        
            &copy; Copyright 2020 Creative Commons Zero v1.0 Universal.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>